### DISCUSSION TITLE
ImGuiFreeType::BuildFontAtlas linking error only in release mode

### DISCUSSION THREAD

**Discussion by Ou7law007** [ocornut/imgui#4720] [New Users Build/Link/Run issues ONLY!]
ImGui Docking v1.86 WIP

Freetype installed using vcpkg.
Both imgui_freetype.h and imgui_freetype.cpp are included in my build.

Calling ImGuiFreeType::BuildFontAtlas() causes the following errors in both x86 and x64 (only in Release mode). The same exact code works in both x86 and x64 in Debug mode
```
(imgui_freetype.obj) : error LNK2001: unresolved external symbol __imp_FT_New_Memory_Face
(imgui_freetype.obj) : error LNK2001: unresolved external symbol __imp_FT_Select_Charmap
(imgui_freetype.obj) : error LNK2001: unresolved external symbol __imp_FT_Done_Face
(imgui_freetype.obj) : error LNK2001: unresolved external symbol __imp_FT_Request_Size
(imgui_freetype.obj) : error LNK2001: unresolved external symbol __imp_FT_Get_Char_Index
(imgui_freetype.obj) : error LNK2001: unresolved external symbol __imp_FT_Load_Glyph
(imgui_freetype.obj) : error LNK2001: unresolved external symbol __imp_FT_GlyphSlot_Embolden
(imgui_freetype.obj) : error LNK2001: unresolved external symbol __imp_FT_GlyphSlot_Oblique
(imgui_freetype.obj) : error LNK2001: unresolved external symbol __imp_FT_Render_Glyph
(imgui_freetype.obj) : error LNK2001: unresolved external symbol __imp_FT_New_Library
(imgui_freetype.obj) : error LNK2001: unresolved external symbol __imp_FT_Add_Default_Modules
(imgui_freetype.obj) : error LNK2001: unresolved external symbol __imp_FT_Done_Library
fatal error LNK1120: 12 unresolved externals
```
I've been trying with this for a couple of days with no success. I tried building Freetype on my own using
<html>
<body>

ft2110.zip | 4M | 19-Jul-2021 16:21
-- | -- | --

<br class="Apple-interchange-newline">
</body>
</html> from https://download.savannah.gnu.org/releases/freetype/, no success.

**Comment by PathogenDavid** [comment]
When you edit project options in Visual Studio by default it only affects the active configuration/platform. Are you sure you configured the linker correctly for the release configuration?

**Reply by Ou7law007** [reply]
Its automatic for vcpkg in Visual Studio. I have a ton of packages automatically linking statically with no problems. Besides, I tried to build it myself and copied the absolute path of the output .lib file with the same result.

**Comment by ocornut** [comment]
Either way the issue is you are not successfully linking Freetype in your project so its not a dear imgui issue, but an issue with your build/VS setup.

**Reply by Ou7law007** [reply]
Alright, I'll keep trying. Thanks

**Comment by Ou7law007** [comment]
It seems like ImGui doesn't support static-freetype. The docs don't mention it either!

**Reply by Ou7law007** [reply]
Dynamically linking all my libs including freetype solved the problem. The triplets workaround also worked fine, hence there's no reason to believe that ImGui supports static linking with (at least the latest version of) freetype.

**Reply by rokups** [reply]
Dear ImGui does not care how you link your libraries. If you are getting linking errors it means you have a mistake in your build configuration somewhere. Error in your initial post indicates that you tried to link to a dynamic library but exported symbols were not found. This would happen if you tried to link to a static freetype library while having `DLL_IMPORT` preprocessor define in build settings. Remove that and it should work.

**Reply by Ou7law007** [reply]
I'm not quite sure what you mean by "you tried to link to a dynamic library", as you can't link to a dynamic library in the first place. All my VS configurations are set to static linkage (Multi-threaded (/MT) and Multi-threaded Debug (/MTd)).

vcpkg does the static linking automatically for all libraries that are installed as static (x64/x86-windows-static) as long as the project configurations are set to static linkage and the vcpkg option "Use Static Libraries" is set to "Yes".

There no DLL_IMPORT anywhere in my project. Apart from that, I tried to manually compile 4 different versions of freetype STATICALLY and got the same result after providing the path to the output freetype.lib.

Apart from ALL that, it literally works in debug mode without changing anything, but not in release mode, so...

Thanks anyways

**Reply by ocornut** [reply]
To use a dynamic library you typically need to link a static library that act as a stub to access functions from the DLL.
We understand you have a linking issue but it doesn't have anything to do with Dear ImGui or imgui_freetype itself. 
Once you fixed your issue would be great to post what you found in case other uses have a similar issue.

**Reply by rokups** [reply]
Actually i misspoke. You tried to link to a static library while your project seems to be configured to link to a dynamic library. Look at freetype source code:

```cpp
#if defined( _WIN32 )

#if defined( FT2_BUILD_LIBRARY ) && defined( DLL_EXPORT )
#define FT_PUBLIC_FUNCTION_ATTRIBUTE  __declspec( dllexport )
#elif defined( DLL_IMPORT )
#define FT_PUBLIC_FUNCTION_ATTRIBUTE  __declspec( dllimport )
#endif
```

When `DLL_IMPORT` it will mark functions with a hint that they are supposed to come from a dynamic library. So you would need to link a library that exports these functions with `__imp_` prefix. That library is what gets produced along side when you compile a DLL and that is what i meant by "link to a dynamic library".

**Reply by Ou7law007** [reply]
I fixed it thanks to @rokups. The fix is bellow.

I've never faced anything like this before! Also, you typically don't have/need direct access to freetype's source code in order to use it in ImGui, if I understand things correctly, you only need the part ImGui exposes to you, hence I did not see the part of source code you posted anywhere before. I'm pretty sure other libraries don't do things like this.

@ocornut, I double checked the tutorial, there's no mention of undefining this macro anywhere.
@rokups Thumbs up to you again. Such a nasty solution, though.

```
// imgui_freetype.cpp
#undef DLL_IMPORT // <-- Added this inside, thanks to rokups
#include "imgui_freetype.h"
#include "imgui_internal.h"     // ImMin,ImMax,ImFontAtlasBuild*,
#include <stdint.h>
#include <ft2build.h>
#include FT_FREETYPE_H          // <freetype/freetype.h>
#include FT_MODULE_H            // <freetype/ftmodapi.h>
#include FT_GLYPH_H             // <freetype/ftglyph.h>
#include FT_SYNTHESIS_H         // <freetype/ftsynth.h>
```

**Comment by ocornut** [comment]
You shouldn’t have to do that though, it probably works for you as it is undoing the opposite setting from your project.

**Reply by Ou7law007** [reply]
I'm unable to find this macro defined anywhere, not even inside my project configurations' preprocessor.
